src/compiler/checker.ts

```diff
@@ -5027,7 +5027,7 @@ namespace ts {
                 }
                 const typeArguments = map(node.typeArguments, getTypeFromTypeNode);
                 const id = getTypeListId(typeArguments);
-                return links.instantiations[id] || (links.instantiations[id] = instantiateType(type, createTypeMapper(typeParameters, typeArguments)));
+                return getOrUpdate(links.instantiations, id, () => instantiateType(type, createTypeMapper(typeParameters, typeArguments)));

```

Comment:
This introduces a new closure and environment record for every call to `getTypeFromTypeAliasReference`. The original approach does not.